File: src/main/java/com/netsensia/rivalchess/model/helper/SliderDirection.java
package com.netsensia.rivalchess.model.helper;

import com.netsensia.rivalchess.model.Piece;
import com.netsensia.rivalchess.model.exception.MoveGenerationException;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public enum SliderDirection {
    N (0,-1),
    NE (1,-1),
    E (1,0),
    SE (1,1),
    S (0,1),
    SW (-1,1),
    W (-1,0),
    NW (-1,-1)
    ;

    private final int xIncrement;
    private final int yIncrement;

    private SliderDirection(final int xIncrement, final int yIncrement) {
        this.xIncrement = xIncrement;
        this.yIncrement = yIncrement;
    }

    private static List<SliderDirection> queenDirections = new ArrayList<>(Arrays.asList(
            SliderDirection.NE, SliderDirection.NW, SliderDirection.SE, SliderDirection.SW,
            SliderDirection.N, SliderDirection.W, SliderDirection.S, SliderDirection.E
    ));

    private static List<SliderDirection> rookDirections = new ArrayList<>(Arrays.asList(
            SliderDirection.N, SliderDirection.W, SliderDirection.S, SliderDirection.E
    ));

    private static List<SliderDirection> bishopDirections = new ArrayList<>(Arrays.asList(
            SliderDirection.NE, SliderDirection.NW, SliderDirection.SE, SliderDirection.SW
    ));

    public int getXIncrement() {
        return this.xIncrement;
    }

    public int getYIncrement() {
        return this.yIncrement;
    }

    public static List<SliderDirection> getDirectionsForPiece(Piece piece) {
        switch (piece) {
            case KING:
            case QUEEN:
                return queenDirections;
            case BISHOP:
                return bishopDirections;
            case ROOK:
                return rookDirections;
            default:
                throw new MoveGenerationException("Can't get directions for a non-sliding piece");
        }
    }
}

File: src/main/java/com/netsensia/rivalchess/model/helper/SliderDirection.kt
package com.netsensia.rivalchess.model.helper

import com.netsensia.rivalchess.model.Piece
import com.netsensia.rivalchess.model.exception.MoveGenerationException
import java.util.*

enum class SliderDirection(val xIncrement: Int, val yIncrement: Int) {
    N(0, -1),
    NE(1, -1),
    E(1, 0),
    SE(1, 1),
    S(0, 1),
    SW(-1, 1),
    W(-1, 0),
    NW(-1, -1);

    companion object {
        private val queenDirections: List<SliderDirection> = ArrayList(listOf(NE, NW, SE, SW, N, W, S, E))
        private val rookDirections: List<SliderDirection> = ArrayList(listOf(N, W, S, E))
        private val bishopDirections: List<SliderDirection> = ArrayList(listOf(NE, NW, SE, SW))

        fun getDirectionsForPiece(piece: Piece): List<SliderDirection> {
            return when (piece) {
                Piece.KING, Piece.QUEEN -> queenDirections
                Piece.BISHOP -> bishopDirections
                Piece.ROOK -> rookDirections
                else -> throw MoveGenerationException("Can't get directions for a non-sliding piece")
            }
        }
    }

}